feat(android): New ANRv2 options (#7246)
getsentry/sentry-docs: src/platforms/android/configuration/app-not-respond.mdx

description: "Learn how to turn off or specify ANR."
---

> # ANR implementation details

Reviewer: ```suggestion
## ANR Implementation Details
```
It would be great if you could add an introductory phrase after the heading (Application Not Responding (ANR)) so that it's not 2 headings one right after the other. Alternatively, we can make ANR Implementation Details the heading instead if it makes more sense.

</application>
```

## ApplicationExitInfo (v2)

This approach reads the [ApplicationExitInfo](https://developer.android.com/reference/android/app/ApplicationExitInfo) API on the next app launch
and asynchronously sends ANR events to Sentry for each ANR in the history, enriching only the latest one with breadcrumbs, contexts, tags, etc.

The integration reports ANR events with `mechanism:AppExitInfo`.

<Note>

> In case [ApplicationExitInfo#getTraceInputStream](<https://developer.android.com/reference/android/app/ApplicationExitInfo#getTraceInputStream()>) returns `null`, the SDK no longer reports an ANR event as these events are not actionable without it.
> 

Reviewer: ```suggestion
If [ApplicationExitInfo#getTraceInputStream](<https://developer.android.com/reference/android/app/ApplicationExitInfo#getTraceInputStream()>) returns `null`, the SDK will no longer report an ANR event, since these events won't be actionable without it.

```

</Note>

![ANR](app-not-respond.png)

### Historical ANRs

> By default, the SDK only reports and enriches the latest ANR and only this one counts towards ANR rate. However, there's
> an `setReportHistoricalAnrs` option available in `SentryOptions`, which enables the SDK to report all of the ANRs from the [getHistoricalExitReasons](<https://developer.android.com/reference/android/app/ActivityManager?hl=en#getHistoricalProcessExitReasons(java.lang.String,%20int,%20int)>) list:
> 

Reviewer: ```suggestion
By default, the SDK only reports and enriches the latest ANR and it's the only one counted towards the ANR rate. However, there's also a `setReportHistoricalAnrs` option available in `SentryOptions`, which enables the SDK to report all ANRs from the [getHistoricalExitReasons](<https://developer.android.com/referL}"pt)>) list:

```

```kotlin
SentryAndroid.init(context) { options ->
  options.isReportHistoricalAnrs = true
}
```

```java
SentryAndroid.init(context) { options ->
  options.setReportHistoricalAnrs(true)
}
```

> This option is useful when updating the SDK to the version where the new ANR implementation has been introduced, to report
> all ANRs happened prior to the SDK update. After that, the SDK will always pick up the latest ANR from the historical exit
> reasons list on next app restart, so there will be no historical ANRs to report.

Reviewer: ```suggestion
This option is useful if you want to update the SDK to the version where the new ANR implementation was introduced, and for reporting all ANRs that took place prior to the SDK update. Other than that, the SDK will always pick up the latest ANR from the historical exit reasons list on next app restart, and there won't be any historical ANRs to report.
```

The integration reports ANR events with `mechanism:HistoricalAppExitInfo`.

### Attaching Thread Dump

> The SDK provides a capability to send ANR thread dump from [ApplicationExitInfo#getTraceInputStream](<https://developer.android.com/reference/android/app/ApplicationExitInfo#getTraceInputStream()>) as attachment. This might be useful to perform
> deeper investigations based on all available information from the OS (in addition to the SDK parsing the thread dump into
> threads with stacktraces):

Reviewer: ```suggestion
The SDK makes it possible to send the ANR thread dump from [ApplicationExitInfo#getTraceInputStream](<https://developer.android.com/reference/android/app/ApplicationExitInfo#getTraceInputStream()>) as an attachment. This is useful for performing deeper investigations using all available information from the OS (in addition to the SDK parsing the thread dump into threads with stack traces):
 ```

```kotlin
SentryAndroid.init(context) { options ->
  options.isAttachAnrThreadDump = true
}
```

```java
SentryAndroid.init(context) { options ->
  options.setAttachAnrThreadDump(true)
}
```

![ANR Thread Dump](app-not-respond-thread-dump.png)

> # ANR Root Cause Analysis

Reviewer: ```suggestion
## ANR Root Cause Analysis
```
